streams/src/main/java/org/apache/kafka/streams/state/internals/RocksDBStore.java

```diff
@@ -505,6 +506,14 @@ private void closeOpenIterators() {
         }
     }
 
+    private ByteBuffer createDirectByteBufferAndPut(byte[] bytes) {
+        ByteBuffer directBuffer = ByteBuffer.allocateDirect(bytes.length);
```

Comment:
I'm not very familiar with the direct buffer usage pattern, but currently it seems we would still try to allocate a new buffer for each put call, whereas I "thought" the main benefits come from reusing the buffer across multiple put calls? @vamossagar12 @ableegoldman @cadonna please correct me if I'm wrong.

Reply:
Yes, I had the same thought as @guozhangwang, but I am also not familiar with direct buffers.

Reply:
The [javadocs](https://docs.oracle.com/javase/8/docs/api/java/nio/ByteBuffer.html) say

> The buffers returned by this method typically have somewhat higher allocation and deallocation costs than non-direct buffers.

So it seems allocating a direct buffer during each put might decrease performance. The internal benchmarks we run on this PR did not show any increase in throughput, but rather a decrease. However, I do not think the decrease was significant, so the cause could also just be a bad day of the environment. Anyways, continuously allocating a direct buffer does not seem to be a good idea, as the javadocs also say:

> It is therefore recommended that direct buffers be allocated primarily for large, long-lived buffers that are subject to the underlying system's native I/O operations. In general it is best to allocate direct buffers only when they yield a measureable gain in program performance. 
   
Another thing to consider:

> The contents of direct buffers may reside outside of the normal garbage-collected heap, and so their impact upon the memory footprint of an application might not be obvious.

Reply:
Thank you @guozhangwang , @cadonna . I think creating it every time does not make much sense. I should have been more careful before adding and asking for the internal benchmarks. In that case, would it even make sense to have it in an API like put and instead use it for putAll()/range/reverseRange/prefixSeek operations? 

That's because in the case of put, it is difficult to know how many put operations may be requested. If users were using the rocksdb library directly, then they can create DirectByteBuffers once and push as many entries as they want.

Based upon my conversations with the rocksdb one of the comments was this:

`Extracting large amounts of data under high concurrency, non-direct byte buffer will bring serious GC problems to the upper level Java services.`

i guess, we can target those APIs? WDYT?

Reply:
The only case I can think of that might have high concurrency on RocksDB state store is with interactive queries. Without interactive queries there is no concurrency on the state stores since only the stream thread that has assigned the stateful task owning the state store accesses the state store.

Reply:
ok. That makes sense. High concurrency is one of the cases where this might be useful. Having said that, on the PR, there are benchmarking numbers for a large number of put operations in a single threaded manner. As per the numbers direct byte buffer was 37% faster and with 0 GC cycles. 

Here is the comment: https://github.com/facebook/rocksdb/pull/2283#issuecomment-561563037

The users of kafka streams might call put() in this manner where in they loop through a bunch of records and use put() to insert. From the state store side, either we create 1 DirectByteBuffer object for put() and keep reusing it- subject to testing. But that might not always be the case.

Reply:
Yes, I think you should try to benchmarks putAll()/range/reverseRange/prefixSeek operations as you proposed with a simple Kafka Streams app. That would be great to better understand the potential of direct buffers for Kafka Streams. Maybe experiment also with different key and value sizes.

Reply:
@vamossagar12 great to see you're about to add direct buffer support, I was actually also looking into this. Just a few comments: as pointed out already, one important aspect to keep in mind with direct buffers is that allocation and disposal is more expensive than with regular heap buffers, so it's important to re-use them, it's common practice to keep a pool of direct buffers around and re-use them on a per-call basis (most I/O libraries to this, e.g., netty, Crail, etc.).  You can keep queues of free and used buffers and move them around during state store operations.

Reply:
Also, we should try to avoid serializing data into byte[] arrays and then copy the data into directBuffers. Instead we should serialize directly into "direct" ByteBuffers. For this we might need to have RocksDBStore implement a ByteBuffer interface, e.g., KeyValueStore<Bytes, ByteBuffer>, or anything similar...

Reply:
> @vamossagar12 great to see you're about to add direct buffer support, I was actually also looking into this. Just a few comments: as pointed out already, one important aspect to keep in mind with direct buffers is that allocation and disposal is more expensive than with regular heap buffers, so it's important to re-use them, it's common practice to keep a pool of direct buffers around and re-use them on a per-call basis (most I/O libraries to this, e.g., netty, Crail, etc.). You can keep queues of free and used buffers and move them around during state store operations.

Thanks @patrickstuedi , that's a great suggestion. I think it makes sense given the expense in allocation/disposal. Couple of questions there:

1)  I think the factor to consider there would be how to manage the pool of used/free buffers? As already pointed by @cadonna  above, the only case where we can expect concurrent access to state stores would be for IQ and otherwise it's single threaded. For the single threaded case, it's going to be straight forward but for IQ, what happens if a request comes in and none of the direct byte buffers are `free` ? Do we re-allocate on the fly?
2) This might be a small issue but calling it out. We need to allocate some capacity to the direct byte buffer. Currently, I am iterating over the list of keys and values and assigning the max length- which again mayn't be the most efficient way. If I were to create the pool, then I would need to know the capacity to allocate upfront. Can this be exposed as a config for the users in that case?

Reply:
> Also, we should try to avoid serializing data into byte[] arrays and then copy the data into directBuffers. Instead we should serialize directly into "direct" ByteBuffers. For this we might need to have RocksDBStore implement a ByteBuffer interface, e.g., KeyValueStore<Bytes, ByteBuffer>, or anything similar...

Yeah.. I did play around with that idea, but the idea was to not change the public APIs. If that makes sense and I can do another round of benchmarking for that, then probably, we will need a KIP.

Reply:
Re (1) if all use cases are single threaded then yes we can allocate some buffer(s) as part of the store. Otherwise, if you need to support multiple concurrent ops then you could pre-populate a queue with a N buffers, and N becomes the maximum number of concurrent requests you can server. If you're queue is empty then a request would have to wait until at least one of the outstanding requests completes and add its buffer to the queue. Again that might all not be needed given the API is single-threaded. 
 
Re (2), is there a max-size, maybe given by the maximum Kafka message size that is configured (if such a limit exists and is not too big)? 

If we don't want to change the API (I guess it would be the RocksDBStore interface that would be changed which is not exposed I think, but still) then splitting this work into part I where we copy heap to direct buffers, and then a part II where we directly serialize into direct buffers is a way to go.

Reply:
I see. I think barring Interactive Queries, the other accesses are single threaded. 
Regarding the max-size, TBH i am not aware. 
Also, regarding changing the API, yeah rocksdb store isn't exposed so the new apis will have to bein KVStore with default implementation.

Reply:
@patrickstuedi , I resorted to creating only 1 instance of DirectByteBuffer(one for putAll keys and values and one for range. I have checked in that code for reference. However, I don't see much of a difference in terms of throughput numbers. I ran it on the same setup(pushing 100 messages on a single partition topic and for each message, push 1M keys using putAll and read through range() call).

I did 3 runs per approach (original v/s ByteBuffer) Here are the numbers:

```
ByteBuffer
Operator: putAll, Real: 87.206 CPU: 371.460 GC: 11.048 GCCount: 30 avg throughput: 1206503.194 op/s p95 throughput: 1423370.427 op/s p99 throughput: 1437098.797 op/s 
Operator: range, Real: 115.121 CPU: 328.090 GC: 7.925 GCCount: 20 avg throughput: 923332.052 op/s p95 throughput: 1261332.193 op/s p99 throughput: 1308671.917 op/s 

Operator: putAll, Real: 87.274 CPU: 382.920 GC: 11.477 GCCount: 29 avg throughput: 1211901.123 op/s p95 throughput: 1432304.571 op/s p99 throughput: 1447671.672 op/s 
Operator: range, Real: 116.886 CPU: 335.610 GC: 8.230 GCCount: 21 avg throughput: 911159.182 op/s p95 throughput: 1234183.861 op/s p99 throughput: 1255574.938 op/s

putAll, Real: 84.438 CPU: 366.390 GC: 10.992 GCCount: 29 avg throughput: 1254820.856 op/s p95 throughput: 1481654.517 op/s p99 throughput: 1508043.762 op/s 
Operator: range, Real: 114.877 CPU: 338.090 GC: 8.465 GCCount: 21 avg throughput: 935155.065 op/s p95 throughput: 1254738.341 op/s p99 throughput: 1280340.688 op/s


Original
Operator: putAll, Real: 95.037 CPU: 378.850 GC: 11.078 GCCount: 29 avg throughput: 1100406.576 op/s p95 throughput: 1292393.480 op/s p99 throughput: 1297963.396 op/s 
Operator: range, Real: 111.177 CPU: 328.180 GC: 8.299 GCCount: 21 avg throughput: 967757.168 op/s p95 throughput: 1281079.326 op/s p99 throughput: 1296172.722 op/s

Operator: putAll, Real: 95.186 CPU: 356.040 GC: 10.132 GCCount: 28 avg throughput: 1092794.645 op/s p95 throughput: 1257861.812 op/s p99 throughput: 1286016.834 op/s 
Operator: range, Real: 112.568 CPU: 347.350 GC: 9.163 GCCount: 25 avg throughput: 952179.810 op/s p95 throughput: 1298717.792 op/s p99 throughput: 1323234.359 op/s 

Operator: putAll, Real: 97.332 CPU: 400.690 GC: 12.000 GCCount: 30 avg throughput: 1079682.386 op/s p95 throughput: 1283925.766 op/s p99 throughput: 1290499.661 op/s 
Operator: range, Real: 109.653 CPU: 319.020 GC: 7.995 GCCount: 20 avg throughput: 980557.905 op/s p95 throughput: 1298695.144 op/s p99 throughput: 1313645.941 op/s
```

One thing that I am noticing is that range is better with the original implementation. This is contrary to the initial number s I had posted [here](https://github.com/apache/kafka/pull/10798#issuecomment-872398600) which was wrong due to a bug in my implementation back then.

Also, the discussion we had about the capacity allocation rocksdb places a limit of 8MB on keys and 3 GB for values. Either ways, it doesn't recommend storing large keys/values. Wanted to know if there are any configs for keys/values sizes for stores? Don't think I could find it.

Lastlty, with the pre-allocation approach for Interactive queries which allows multi threaded access, do we want to restirct by N as you specified above?

What do you. and (@guozhangwang / @cadonna ) think?